tiledb/sm/filesystem/s3.cc

```diff
@@ -267,38 +268,50 @@ Status S3::disconnect() {
   if (multipart_upload_states_.size() > 0) {
     RETURN_NOT_OK(init_client());
 
-    for (auto& kv : multipart_upload_states_) {
-      const MultiPartUploadState& state = kv.second;
-
-      if (state.st.ok()) {
-        Aws::S3::Model::CompleteMultipartUploadRequest complete_request =
-            make_multipart_complete_request(state);
-        auto outcome = client_->CompleteMultipartUpload(complete_request);
-        if (!outcome.IsSuccess()) {
-          const Status st = LOG_STATUS(Status::S3Error(
-              std::string("Failed to disconnect and flush S3 objects. ") +
-              outcome_error_message(outcome)));
-          if (!st.ok()) {
-            ret_st = st;
-          }
-        }
-      } else {
-        Aws::S3::Model::AbortMultipartUploadRequest abort_request =
-            make_multipart_abort_request(state);
-        auto outcome = client_->AbortMultipartUpload(abort_request);
-        if (!outcome.IsSuccess()) {
-          const Status st = LOG_STATUS(Status::S3Error(
-              std::string("Failed to disconnect and flush S3 objects. ") +
-              outcome_error_message(outcome)));
-          if (!st.ok()) {
-            ret_st = st;
+    std::vector<const MultiPartUploadState*> states;
+    states.reserve(multipart_upload_states_.size());
+    for (auto& kv : multipart_upload_states_)
+      states.emplace_back(&kv.second);
+    multipart_lck.unlock();
+
```

Comment:
There's now a race here with L435. `S3::finish_flush_object` can now take `multipart_lck` and erase the state.

We need to strictly enforce the following lock acquisition ordering to ensure the same level of thread safety:
- lock multipart_upload_mtx_
- lock MultiPartState::mtx
- unlock in any order

We have similar races elsewhere so I won't mention them, but please apply this ordering elsewhere. For this specific scenario, I would modify states to be of type `std::vector<std::pair<const MultiPartUploadState*, std::unique_lock<std::mutex>>` and store their held lock alongside them in a pair.